File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
n, need_height = map(int, input().split())  # 나무의 수, 필요한 나무의 길이
wood_heights = list(map(int, input().split()))  # 나무들의 높이

# 이분탐색
start = 0
end = max(wood_heights)
mid = (start + end) // 2

result_height = 0

while start <= end:
    result_height = 0
    mid = (start + end) // 2
    result_height = sum([wood_height - mid if mid < wood_height else 0 for wood_height in wood_heights])  # 설정된 mid 로 result_height 를 구함


    if result_height >= need_height:
        result = mid
        start = mid + 1
    elif result_height < need_height:
        end = mid - 1

print(result)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.cs
using System;
using System.Text;

namespace Baekjoon {
    class Program {
        static void Main()
        {
            var part = Console.ReadLine().Split();
            int n = int.Parse(part[0]);
            int m = int.Parse(part[1]);

            List<int> heights = Console.ReadLine().Split().Select(int.Parse).ToList();

            Console.WriteLine(CalcMaxHeight(m, heights));
        }

        static int CalcMaxHeight(int wishLength, List<int> heights)
        {

            int low = 0;
            int high = heights.Max();

            // 목표: 원하는 길이만큼 자를 수 있는 최대높이 찾기
            int result = 0;
            while (low <= high)
            {
                int mid = low + ((high - low) / 2);
                
                // Console.WriteLine($"CutSize: {CutSize(heights, mid)}({mid})");
                if (CutSize(heights, mid) >= wishLength) // 원하는 길이보다 같거나 긺
                {
                    // 최대높이 갱신
                    result = mid; 
                    low = mid + 1;
                }
                else // 원하는 길이보다 짧음
                {
                    high = mid - 1;
                }
            }

            return result;
        }

        static long CutSize(List<int> heights, int cutHeight)
        {
            long cutSize = 0;
            foreach (var height in heights)
            {
                if (height > cutHeight)
                    cutSize += (height - cutHeight);
            }

            return cutSize;
        }
    }
}